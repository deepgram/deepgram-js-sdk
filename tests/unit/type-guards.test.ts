import {
  isUrlSource,
  isTextSource,
  isFileSource,
  isLiveSchema,
  isDeepgramClientOptions,
} from "../../src/lib/helpers";
import { Readable } from "stream";

describe("Unit Tests - Type Guards and Validators", () => {
  describe("isUrlSource", () => {
    it("should identify URL sources correctly", () => {
      const testCases = [
        { url: "https://dpgr.am/spacewalk.wav" },
        { text: "This is not a URL source" },
        { url: "file://local.wav" },
        { buffer: Buffer.from("fake audio") },
        { invalidProperty: "value" },
        {},
        null,
        undefined,
      ];

      const results = testCases.map((source) => ({
        input: source,
        output: source ? isUrlSource(source as any) : false,
      }));

      expect(results).toMatchSnapshot();
    });
  });

  describe("isTextSource", () => {
    it("should identify text sources correctly", () => {
      const testCases = [
        { text: "This is a text source" },
        { url: "https://dpgr.am/spacewalk.wav" },
        { text: "" },
        { text: "Multi-line\ntext content" },
        { invalidProperty: "value" },
        {},
        null,
        undefined,
      ];

      const results = testCases.map((source) => ({
        input: source,
        output: source ? isTextSource(source as any) : false,
      }));

      expect(results).toMatchSnapshot();
    });
  });

  describe("isFileSource", () => {
    it("should identify file sources correctly", () => {
      const buffer = Buffer.from("fake audio data");
      const readable = new Readable({
        read() {
          this.push("audio data");
          this.push(null);
        },
      });

      const testCases = [
        buffer,
        readable,
        { url: "https://dpgr.am/spacewalk.wav" },
        { text: "This is text" },
        "string data",
        {},
        null,
        undefined,
      ];

      const results = testCases.map((source) => ({
        input: source?.constructor?.name || typeof source,
        output: isFileSource(source as any),
      }));

      expect(results).toMatchSnapshot();
    });
  });

  describe("isLiveSchema", () => {
    it("should identify live schema correctly", () => {
      const testCases = [
        { interim_results: true },
        { interim_results: false },
        { language: "en-US", model: "nova-2" },
        { interim_results: true, language: "en-US" },
        {},
        null,
        undefined,
        "not an object",
      ];

      const results = testCases.map((arg) => ({
        input: arg,
        output: isLiveSchema(arg),
      }));

      expect(results).toMatchSnapshot();
    });
  });

  describe("isDeepgramClientOptions", () => {
    it("should identify Deepgram client options correctly", () => {
      const testCases = [
        { global: { url: "https://api.deepgram.com" } },
        { global: {} },
        { apiKey: "test-key" },
        { global: { headers: { Custom: "Header" } } },
        {},
        null,
        undefined,
        "not an object",
      ];

      const results = testCases.map((arg) => ({
        input: arg,
        output: isDeepgramClientOptions(arg),
      }));

      expect(results).toMatchSnapshot();
    });
  });
});
