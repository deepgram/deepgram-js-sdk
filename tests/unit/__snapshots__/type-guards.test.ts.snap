// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Tests - Type Guards and Validators isDeepgramClientOptions should identify Deepgram client options correctly 1`] = `
[
  {
    "input": {
      "global": {
        "url": "https://api.deepgram.com",
      },
    },
    "output": true,
  },
  {
    "input": {
      "global": {},
    },
    "output": true,
  },
  {
    "input": {
      "apiKey": "test-key",
    },
    "output": false,
  },
  {
    "input": {
      "global": {
        "headers": {
          "Custom": "Header",
        },
      },
    },
    "output": true,
  },
  {
    "input": {},
    "output": false,
  },
  {
    "input": null,
    "output": null,
  },
  {
    "input": undefined,
    "output": undefined,
  },
  {
    "input": "not an object",
    "output": false,
  },
]
`;

exports[`Unit Tests - Type Guards and Validators isFileSource should identify file sources correctly 1`] = `
[
  {
    "input": "Buffer",
    "output": true,
  },
  {
    "input": "Readable",
    "output": true,
  },
  {
    "input": "Object",
    "output": true,
  },
  {
    "input": "Object",
    "output": true,
  },
  {
    "input": "String",
    "output": true,
  },
  {
    "input": "Object",
    "output": true,
  },
  {
    "input": "object",
    "output": false,
  },
  {
    "input": "undefined",
    "output": false,
  },
]
`;

exports[`Unit Tests - Type Guards and Validators isLiveSchema should identify live schema correctly 1`] = `
[
  {
    "input": {
      "interim_results": true,
    },
    "output": true,
  },
  {
    "input": {
      "interim_results": false,
    },
    "output": true,
  },
  {
    "input": {
      "language": "en-US",
      "model": "nova-2",
    },
    "output": false,
  },
  {
    "input": {
      "interim_results": true,
      "language": "en-US",
    },
    "output": true,
  },
  {
    "input": {},
    "output": false,
  },
  {
    "input": null,
    "output": null,
  },
  {
    "input": undefined,
    "output": undefined,
  },
  {
    "input": "not an object",
    "output": false,
  },
]
`;

exports[`Unit Tests - Type Guards and Validators isTextSource should identify text sources correctly 1`] = `
[
  {
    "input": {
      "text": "This is a text source",
    },
    "output": true,
  },
  {
    "input": {
      "url": "https://dpgr.am/spacewalk.wav",
    },
    "output": false,
  },
  {
    "input": {
      "text": "",
    },
    "output": false,
  },
  {
    "input": {
      "text": "Multi-line
text content",
    },
    "output": true,
  },
  {
    "input": {
      "invalidProperty": "value",
    },
    "output": false,
  },
  {
    "input": {},
    "output": false,
  },
  {
    "input": null,
    "output": false,
  },
  {
    "input": undefined,
    "output": false,
  },
]
`;

exports[`Unit Tests - Type Guards and Validators isUrlSource should identify URL sources correctly 1`] = `
[
  {
    "input": {
      "url": "https://dpgr.am/spacewalk.wav",
    },
    "output": true,
  },
  {
    "input": {
      "text": "This is not a URL source",
    },
    "output": false,
  },
  {
    "input": {
      "url": "file://local.wav",
    },
    "output": true,
  },
  {
    "input": {
      "buffer": {
        "data": [
          102,
          97,
          107,
          101,
          32,
          97,
          117,
          100,
          105,
          111,
        ],
        "type": "Buffer",
      },
    },
    "output": false,
  },
  {
    "input": {
      "invalidProperty": "value",
    },
    "output": false,
  },
  {
    "input": {},
    "output": false,
  },
  {
    "input": null,
    "output": false,
  },
  {
    "input": undefined,
    "output": false,
  },
]
`;
